hooks/tk-multi-publish2/photoshopcc.basic/upload_version.py

```diff
@@ -0,0 +1,348 @@
+                    "text": "<pre>%s</pre>" % (
+                    pprint.pformat(version_data),)
+                }
+            }
+        )
+
+        # create the version
+        self.logger.info("Creating version for review...")
+        version = self.parent.shotgun.create("Version", version_data)
+
+        # stash the version info in the item just in case
+        item.properties["sg_version_data"] = version
+
+        # upload the file to SG
+        self.logger.info("Uploading content...")
+        self.parent.shotgun.upload(
+            "Version",
+            version["id"],
+            upload_path,
+            "sg_uploaded_movie"
+        )
+        self.logger.info("Upload complete!")
+
+        # go ahead and update the publish thumbnail (if there was one)
+        if publish_data:
+            self.logger.info("Updating publish thumbnail...")
+            self.parent.shotgun.upload_thumbnail(
+                publish_data["type"],
+                publish_data["id"],
+                upload_path
+            )
+            self.logger.info("Publish thumbnail updated!")
+
+        item.properties["upload_path"] = upload_path
+
+    def finalize(self, settings, item):
+        """
+        Execute the finalization pass. This pass executes once all the publish
+        tasks have completed, and can for example be used to version up files.
+
+        :param settings: Dictionary of Settings. The keys are strings, matching
+            the keys returned in the settings property. The values are `Setting`
+            instances.
+        :param item: Item to process
+        """
+
+        version = item.properties["sg_version_data"]
+
+        self.logger.info(
+            "Version uploaded for Photoshop document",
+            extra={
+                "action_show_in_shotgun": {
+                    "label": "Show Version",
+                    "tooltip": "Reveal the version in Shotgun.",
+                    "entity": version
+                }
+            }
+        )
+
+        upload_path = item.properties["upload_path"]
+
+        # remove the tmp file
+        try:
+            os.remove(upload_path)
+        except Exception:
+            self.logger.warn("Unable to remove temp file: %s" % (upload_path,))
+            pass
+
+    def _get_version_entity(self, item):
+        """
+        Returns the best entity to link the version to.
+        """
+
+        if item.context.entity:
+            return item.context.entity
+        elif item.context.project:
+            return item.context.project
+        else:
+            return None
+
+
+def _get_save_action(document):
+    """
+    Simple helper for returning a log action dict for saving the session
+    """
+    return {
+        "action_button": {
+            "label": "Save",
+            "tooltip": "Save the current session",
+            "callback": lambda: document.save()
+        }
+    }
+
+
+def _document_path(document):
```

Comment:
Good example of why this makes sense to live in `AdobeBridge` since it's duplicated in multiple modules.